raiden/network/proxies/token_network.py

```diff
@@ -2175,73 +2271,171 @@ def settle(
+                    block_identifier=failed_at_blockhash,
                     channel_identifier=channel_identifier,
                 )
-                raise RaidenUnrecoverableError(f"{error_prefix}. {msg}")
+                our_details = self._detail_participant(
+                    channel_identifier=channel_identifier,
+                    detail_for=self.node_address,
+                    partner=partner,
+                    block_identifier=failed_at_blockhash,
+                )
+                partner_details = self._detail_participant(
+                    channel_identifier=channel_identifier,
+                    detail_for=partner,
+                    partner=self.node_address,
+                    block_identifier=failed_at_blockhash,
+                )
+
+                our_balance_hash = hash_balance_data(
+                    transferred_amount=transferred_amount,
+                    locked_amount=locked_amount,
+                    locksroot=locksroot,
+                )
+                partner_balance_hash = hash_balance_data(
+                    transferred_amount=partner_transferred_amount,
+                    locked_amount=partner_locked_amount,
+                    locksroot=partner_locksroot,
+                )
+
+                if channel_identifier != channel_onchain_detail.channel_identifier:
+                    msg = (
+                        f"The provided channel identifier {channel_identifier} "
+                        f"does not match onchain channel_identifier "
+                        f"{channel_onchain_detail.channel_identifier}."
+                    )
+                    raise RaidenRecoverableError(msg)
+
+                if channel_onchain_detail.state == ChannelState.SETTLED:
+                    raise RaidenRecoverableError("Channel is already settled")
+                elif channel_onchain_detail.state == ChannelState.REMOVED:
+                    raise RaidenRecoverableError(
+                        "Channel is already unlocked. It cannot be settled"
+                    )
+                elif channel_onchain_detail.state == ChannelState.OPENED:
+                    raise RaidenUnrecoverableError("Channel is still open. It cannot be settled")
+                elif channel_onchain_detail.state == ChannelState.CLOSED:
+                    if failed_at_blocknumber < channel_onchain_detail.settle_block_number:
+                        raise RaidenUnrecoverableError(
+                            "Channel cannot be settled before settlement window is over"
+                        )
+
+                # on-chain balance hashes are never updated after
+                # the channel was closed / update transfer was called.
+                # Therefore, a non-matching balance hash values
+                # mean that the channel was removed / a new
+                # channel was created.
+                if our_details.balance_hash != our_balance_hash:
```

Comment:
Okay, so with the change I suggested above we know at this point that this channel exists, and it is at the state closed, so if the balance_hash is different then we have an unrecoverable error.